tensorflow_addons/optimizers/utils.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Additional Utilities used for tfa.optimizers."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+def fit_bn(model, *args, **kwargs):
+    """Resets batch normalization layers of model, and recalculates the
+    statistics for each batchnorm layer by running a pass on the data.
+
+    Args:
+        model: An instance of tf.keras.Model
+        *args, **kwargs: Params that'll be passed to `.fit` method of model
+    """
+    kwargs['epochs'] = 1
+    if not isinstance(model, tf.keras.Model):
+        raise TypeError("model must be an instance of tf.keras.Model")
+
+    if not model.built:
+        raise ValueError("Call `fit_bn` after the model is built and trained")
+
+    assign_ops = []
+    for layer in model.layers:
+        if isinstance(layer, tf.keras.layers.BatchNormalization):
+            if not layer.built:
+                continue
+
+            assign_ops.append([
+                layer.moving_mean.assign(tf.zeros_like(layer.moving_mean)),
+                layer.moving_variance.assign(
+                    tf.ones_like(layer.moving_variance))
+            ])
+
+    assign_op = tf.group(assign_ops)
+
+    _trainable = model.trainable
+    _metrics = model._metrics
+    model.trainable = False
```

Comment:
BatchNormalization works a little different. `moving_mean` and `moving_variance` aren't trainable variables. So when we do `trainable=False`, they can still be changed during fit. 

I printed the variables before and after `fit` to make sure they are being updated.

[Docs](https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization?version=nightly)

Reply:
I don't actually mean the moving_mean and moving_average terms. Sorry for not making my comment clear.

Batch norm does have two trainable parameters which are the `gamma` and the `beta` terms which might need to change since the weights have changed and the moving_mean and moving_variance preceding them is changing as well.

The `gamma` and `beta` terms are canonically used to scale and shift the batch normalization input after the use of the moving_mean and moving_variance.

[Link](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/layers/normalization.py#L174) to Keras' call to the backend for a reference for `beta` and `gamma`. Tensorflow [doc](https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization?version=nightly) for batch normalization.

Reply:
`beta` and `gamma` should not be changed as per [the implementation](https://github.com/timgaripov/swa/blob/4a2ddfdb2692eda91f2ac41533b62027976c605b/utils.py) of the paper's authors. They never do a backprop on the network (which is when `beta` and `gamma` are updated). 

> Batch norm does have two trainable parameters, the gamma and the beta terms which might need to change since the weights have changed and the moving_mean and moving_variance preceding them are changing as well.

The distribution of the activations (before scaling using `gamma` and `beta`) should (_may?_) be the same between using the original weights and averaged weights as long as `moving_mean` and `moving_variance` are updated. At least this is the assumption made when fine-tuning batch norm on new data, and I think the logic carries over to this case as well. 

Also, as soon as `beta` and `gamma` are changed, then all the following layers need to be updated as well. That boils down to another epoch of full network training. 

Thanks for taking the time to do this. Let me know if what I've understood is not correct.

Reply:
Yep. It seems like it `beta` and `gamma` don't change in the author's original implementation and yeah I guess if you think of this as fine-tuning batch norm then this would make sense.

I felt like there is a large enough change in `beta` and `gamma` to allow them to be learnable (without resetting) but I have no evidence to back it up.

Everything looks good otherwise. I'll resolve this comment.